proposals/025-lambda-1.md

```diff
@@ -0,0 +1,141 @@
+
+## Proposal
+
+The proposal is to add the following syntax for lambda expressions:
+```slang
+(parameter_list) => expression
+```
+
+or
+
+```slang
+(parameter_list) => { statement_list }
+```
+
+Where `parameter_list` is a comma-separated list of parameters, same as those in ordinary functions,
+and `expression` or `statement_list` defines the body of the lambda. For examples, these two lambdas
+achieve similar results:
+
+```slang
+(int x) => return x > 0 ? x : 0
+
+(int x) => {
+    if (x > 0) {
+        return x;
+    } else {
+        return 0;
+    }
+}
+```
+
+A lambda expression will evaluate to an annoymous struct type that implements the `IFunc` interface
+during type checking. The return type of the lambda function is determined by the body expression (in the case
+of the lambda expression contains a simple expression body), or the value of the first return statement in the
+case of a statement body. If the lambda function body contains more than one return statements, and the result
+value specified in any of the return statements cannot coerce to the result type of the first return statement,
+the compiler will diagnose a type mismatch error on the mismatching return statement.
+
+Lambda expressions can be used in positions that accepts an `IFunc`:
+
+```
+void apply(IFunc<int, float> x) {...}
+
+void test()
+{
+    apply((float x)=>(int)x+1); // OK, passing lambda to `IFunc<int, float>`.
+}
+```
+
+## Translation
+
+Immutable lambda expressions translates into a struct type implementing the corresponding `IFunc` interface.
+For example, given the following code:
+
+```slang
+void test()
+{
+    int c = 0;
+    let lam = (int x) => x + c;
+    int d = lam(2);
+}
+```
+
+The compiler will translate it into:
+
+```slang
+void test()
+{
+    int c = 0;
+    struct _slang_Lambda_test_0 : IFunc<int, int> {
+        int c;
+        __init(int in_c) {
+            c = in_c;
+        }
+        int operator()(int x) {
+            return x + c;
+        }
+    }
+    let lam = _slang_Lambda_test_0(c);
+    int d = lam.operator()(2);
+}
+```
+
+## Restrictions
+
+Lambda expression in this proposed version can only read captured variables, but not modify them.
+For example:
+
+```slang
+void test()
+{
+    int c = 0;
+    let lam = (int x) {
+        c = c + 1; // Error: c is read-only here.
+        return x + c;
+    };
+    int d = lam(3);
+}
+```
+
+We plan to allow mutating captured variables in a future proposal.
+
+Lambda expression is not allowed to have mutable parameters, such as `inout` or `out` parameters in this version.
+
+A variable whose type is `[NonCopyable]` cannot be captured in a lambda expression.
+
+# Conclusion
+
+This proposal adds limited support for lambda expressions that cannot mutate its captured environment.
+Although being limited in functionality, this kind of lambda expressions will still be very useful
+in many scenarios including the cooperative-matrix operations.
```

Comment:
One thing I might have expected to see (although I haven't looked at the relevant SPIR-V extensions, so maybe they can handle something like a closure...) is a distinction being made in the type system between "stateless" closures (more or less equivalent to just a function pointer, with the given signature) vs. "stateful" ones (more or less some hidden storage of type `T` and a pointer to a function that has the given signature plus an extra `T` parameter).

There's a lot of nuance there around how to handle calling conventions and so forth, and around avoiding generating extra "thunk" values when, e.g., passing a plain old function to an operation that accepts/expects a closure).

Reply:
We could potentially express this as:
```
interface IStaticFunc<T, each TArgs> {
    static T operator()(expand each TArgs args);
}
```
But this seems like it is worth having a separate proposal for.